### ISSUE TITLE
Early data problem with s_client

### ISSUE THREAD

**Issue by kroeckx** [openssl/openssl#2978]
Hi,

I've been testing early data with s_server/s_client. On the server side I see:
```
ACCEPT
Early data received:
test

End of early data
-----BEGIN SSL SESSION PARAMETERS-----
MGMCAQECAgMEBAITAgQABDCPd6mZMgfXaVPk970I4uJX/8Th9EHKvsAg7LfudQIz
WNXdqxZ+FmAO7iUK4OD6XFKhBgIEWMwnnqIEAgIcIKQGBAQBAAAArgYCBAQDAgGv
BAICQAA=
-----END SSL SESSION PARAMETERS-----
Supported Elliptic Groups: X25519:P-256:P-521:P-384
Shared Elliptic groups: X25519:P-256:P-521:P-384
CIPHER is TLS13-AES-256-GCM-SHA384
Reused session-id
Secure Renegotiation IS supported
```

On the client side I see:
```
New, TLSv1.3, Cipher is TLS13-AES-256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was rejected
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS13-AES-256-GCM-SHA384
    Session-ID: F551DD7821EBED38730B3953FC22E6C74C71186D52E8337DFFFE4D408FA4B33E
    Session-ID-ctx:
    Master-Key: BA06DBD3A9E188670335F14A851228EB81777EEA92BA2AFB04C11ADA5A7FD1F200E2417477157C7045C5EC2E8E92793F
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 7200 (seconds)
    TLS session ticket:
    0000 - 01 02 03 04 05 06 07 08-09 0a 0b 0c 0d 0e 0f 10   ................
    0010 - 01 02 03 04 05 06 07 08-09 0a 0b 0c 0d 0e 0f 10   ................
    0020 - 2f af f9 73 2b f0 cb 67-b9 94 c8 22 83 0f 56 09   /..s+..g..."..V.
    0030 - e3 59 49 df e0 af dc b5-5b 2c b2 49 6a c4 42 27   .YI.....[,.Ij.B'
    0040 - 8a 3b 29 e4 c0 02 4b d2-dc a4 ec 79 fb a8 d8 74   .;)...K....y...t
    0050 - ba fe 18 86 f6 33 f6 d9-dc cf 7c 90 d4 c6 19 13   .....3....|.....
    0060 - 84 f2 91 6b c7 de 52 dc-df b6 d2 c3 a2 a0 8b 07   ...k..R.........
    0070 - 8c 46 9f 1e 6b af d5 17-04 0c fa 3b d4 12 3a a7   .F..k......;..:.
    0080 - cc 33 02 51 22 7a 83 e4-66 ce 47 27 d5 d4 96 ca   .3.Q"z..f.G'....
    0090 - 27 30 36 7e 60 c3 1d 91-d2 e6 dd 53 50 68 54 be   '06~`......SPhT.
    00a0 - 79 2f d4 70 f8 bc 38 44-d9 30 f5 bf 9f 39 e5 65   y/.p..8D.0...9.e

    Start Time: 1489774207
    Timeout   : 7200 (sec)
    Verify return code: 21 (unable to verify the first certificate)
    Extended master secret: no
```

I'm confused that the server shows it received the early data while the client says that it was rejected.

**Comment by kaduk** [comment]
I am seeing something similar; testing against https://tls13.crypto.mozilla.org/ the server is clearly receiving my early data (which contains a basic GET), since my regular data is blank and it returns a page to me, which it does not do when I just feed it blank data.  But I still see the "Early data was rejected" message from s_client.  So, maybe it's just a reporting issue in s_client...

**Comment by kaduk** [comment]
Well, I guess the behavior I observe would also be consistent with s_client silently replaying the 0-RTT data as 1-RTT data after the server rejects early data, but I really hope we're not doing that...

**Comment by kaduk** [comment]
Yeah, this is just a reporting issue.  Running in gdb, we print out "Early data was rejected" before s_client even gets to tls_parse_stoc_early_data() (which sets it to SSL_EARLY_DATA_ACCEPTED), so the print statement is just printing the value it was initialized with when we sent the early data.

**Comment by mattcaswell** [comment]
Yes, I spotted this the other day...and then forgot all about it again! Sorry. Its an s_client bug. Basically s_client gets a bit confused and thinks its finished connecting, and prints out the connection summary information too early (before we know that the early_data has been accepted).

**Comment by mattcaswell** [comment]
This got fixed by commit 12557a344.
